client/concordclient/include/client/concordclient/remote_update_data.hpp

```diff
@@ -46,6 +46,11 @@ struct Update {
   std::string span_context;
 };
 
-typedef std::variant<Update, EventGroup> EventVariant;
+struct SnapshotKVPair {
+  std::string key;
+  std::string val;
+};
+
+typedef std::variant<Update, EventGroup, SnapshotKVPair> RemoteData;
```

Comment:
Please **do not** mix any data structures with what is used for the TRC.
The only shared code between the snapshot client and the TRC should be the gRPC connection.
Concord client will initialize both objects separately and expose their APIs separately. Meaning, the types should be separate as well. Otherwise, this will become a maintenance burden if we change one or the other.

If you choose to use the same update mechanism then I suggest to copy the types and adjust them to the state snapshot needs.
However, I was under the impression that we can simply forward the incoming data from the replicas and hand it over to the clientservice via a stream object without even looking at the data. Why do we need `SnapshotKVPair` in the first place?

Also, the TRC and the snapshot client would never share the same queue. Meaning, there is no need to have a distinction between the types of the elements at runtime.

Reply:
Thanks @robem for setting this discussion. My view is a little different here. I want to explain my reasoning for this using the concept of Type Subset.
Consider that I am building a container (like vector) for some reason. Normally all the STL containers are created for all kinds of types. These containers are close. So we can create compositions of containers of any type once we have a generic container. These containers are generally created as template,
template<typename T>
class Container{...};
So above the container is a generic container.
But sometimes for some reason (which i don't know in this case), containers are created for a subset of types. In such cases the container is not for everything. There are many ways to define such a container. One way is to define the container on top of a variant as below:
typedef variant<T1, T2> DataType;
class Container {
DataType dt;
...
};
Now, if for some valid reason, this Container needs to use the third type then adding the same in the variant is needed as this Container is designed on the subset of the type.
For Inter Thread Communication (ITC) we have a subset type container called BasicUpdateQueue. This container is exactly required for the ITC. So this data type is used here.
Also, every snapshot request is creating a new queue and provides the same to read the snapshot, so the queue object is not shared with TRC and Snapshot. Its the Container that is shared.

Reply:
Thanks, @robem for this comment. I have removed the type subsetting in the container and then removed SnapshotKVPair from EventVariant. Now everything is consistent.